CDCgov/prime-reportstream
# Router Refactor 1: Install `react-router-dom` v6 and refactor main app router

> This is one part of a larger refactor. This PR merges into a shadow-master branch: `experience/shadow/router-refactor`

This PR makes a few changes in the adoption of react-router v6:

- AuthRoute pattern is no longer supported; now use `AuthElement` to authenticate a user before loading a page.
- Deprecation of `useHistory` and `RouteComponentProps`; now use `useNavigate` and `useParams`. (**NOTE**: We can now more easily throw Errors if any of the params come up undefined -- great perk of the pattern)

Additionally, some refactors lead to added capabilities:

- `AuthElement` supports one or many `MemberType` arguments for auth
- `AuthElement` supports a feature flag
- Throws on components expecting url params but receiving undefined (negates the need for undefined checks in code, AND plays into our error boundary tech-debt)

Test Steps:
1. Pull this branch and run `yarn` to install new packages
2. Start the front-end
3. The home page should load. That is the extend of the work in _this_ PR.

## Changes
- Pages requiring auth now all export `[name]WithAuth` for use in routing
- `AppRouter` extracted from `App` and refactored to meet v6 requirements

## Checklist

### Testing

### Process

## Linked Issues
- Fixes #issue

## To Be Done
*Create GitHub issues to track the work remaining, if any*
- #issue

## Specific Security-related subjects a reviewer should pay specific attention to
- Does this PR introduce new endpoints?
    - new endpoint A
    - new endpoint B
- Does this PR include changes in authentication and/or authorization of existing endpoints?
- Does this change introduce new dependencies that need vetting?
- Does this change require changes to our infrastructure?
- Does logging contain sensitive data?
- Does this PR include or remove any sensitive information itself?

If you answered '_yes_' to any of the questions above, conduct a detailed Review that addresses at least:

- What are the potential security threats and mitigations? Please list the _STRIDE_ threats and how they are mitigated
    - **S**poofing (faking authenticity)
        - Threat _T_, which could be achieved by _A_, is mitigated by _M_
    - **T**ampering (influence or sabotage the integrity of information, data, or system)
    - **R**epudiation (the ability to dispute the origin or originator of an action)
    - **I**nformation disclosure (data made available to entities who should not have it)
    - **D**enial of service (make a resource unavailable)
    - **E**levation of Privilege (reduce restrictions that apply or gain privileges one should not have)
- Have you ensured logging does not contain sensitive data?
- Have you received any additional approvals needed for this change?
